src/core/lib/security/credentials/google_default/google_default_credentials.cc

```diff
@@ -273,59 +274,82 @@ static grpc_error* create_default_creds_from_path(
   return error;
 }
 
-grpc_channel_credentials* grpc_google_default_credentials_create() {
-  grpc_channel_credentials* result = nullptr;
+static void update_tenancy() {
+  gpr_once_init(&g_once, init_default_credentials);
+  gpr_mu_lock(&g_state_mu);
+
+  /* Try a platform-provided hint for GCE. */
+  if (!g_metadata_server_available) {
+    g_metadata_server_available = g_gce_tenancy_checker();
+  }
+  /* TODO: Add a platform-provided hint for GAE. */
+
+  /* Do a network test for metadata server. */
+  if (!g_metadata_server_available) {
+    g_metadata_server_available = is_metadata_server_reachable();
+  }
+  gpr_mu_unlock(&g_state_mu);
+}
+
+static bool metadata_server_available() {
+  bool available = false;
```

Comment:
This could be written as:

```
grpc_core::MutexLock lock(&g_state_mu);
return static_cast<bool>(g_metadata_server_available);
```

Reply:
I didn't realize we had our own version of `std::unique_lock`. Neat.